PYIC-5872: Expose core's internal API in dev

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Expose core's internal API to requets with an API key.

### Why did it change

[PYIC-5872: Expose internal API in dev](https://github.com/govuk-one-login/ipv-core-back/pull/1966/commits/796c7896128268d063e020c4b9cc0bd1ba7ec859)

This change allows core's internal API to be accessed with an API key
when deployed to either of the dev accounts.

It adds a usage plan with API key to the internal API, and then makes
use of an API key required only for the dev accounts. It also adds DNS
to the API gateway to make use of the internal API easier.

Also, to allow programatic access to the API key, a new resource policy
is created to allow the OIDC identity provider enough permissions to
fetch the API key.

It removes the `CoreFrontLocal` parameter in favour of defaulting to
this behavour for dev accounts. This is because changing back to a
private API from the new regional version is a two step process - so not
as easy as just toggling the parameter.

It's a two step process as you can't change from regional to private if
it has custom domains associated with it. In CloudFormation there isn't
a way to specify that something should be removed before making changes,
resulting in having to remove the custom domain before changing the API
to private.

Ideally I would have liked to have conditionally created the UsagePlan -
only for the dev accounts. However there appears to be a bug in cfn-lint
that means we can't use an If function with the CreateUsagePlan
property.

The GHA workflow included in this PR is just to demonstrate that the
gateway is now available with an API key, and also how to
programatically access the API key. This would not be merged with the
PR.

[PYIC-5872: Don't use SAM UsagePlan](https://github.com/govuk-one-login/ipv-core-back/pull/1966/commits/e879e14f808012d0d299768a5a1d826f48ff7e13)

SAM allows an easy way to create a usage plan and API key, by defining
the `UsagePlan` property in an `Api` resource. Unfortunately you can't
use an `!If` function with it do conditionally create it.

This removes the use of it and defines the resources requied using
plain old CloudFormation. This allows us to only create them when we
want.

Using the previous method would have caused a usage plan and api key to
be created in all envs including prod. This felt like murky ground - the
non dev api's would still be private so it wouldn't have been an issue,
but feels safer to only create them when actually needed.

[PYIC-5872: Create separate API for internal testing](https://github.com/govuk-one-login/ipv-core-back/pull/1966/commits/5c8765d81953dff79d39bb5e8f680e18054b0976)

This creates a new API gateway to run in parallel with the original.
This means that the VPC protected internal API can still be used by
core-front without any changes, and the new regional, API token
protected API gateway can be used for testing.

cfn-lint has an issue with having an event defined in a lambda that
includes a reference to a conditionally created resource, even if that
resource is behind the same condition. This has lead to the ignore
written into the metadata of a number of resources.

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-5872](https://govukverify.atlassian.net/browse/PYIC-5872)

[PYIC-5872]: https://govukverify.atlassian.net/browse/PYIC-5872?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-5872: Expose internal API in dev

This change allows core's internal API to be accessed with an API key
when deployed to either of the dev accounts.

It adds a usage plan with API key to the internal API, and then makes
use of an API key required only for the dev accounts. It also adds DNS
to the API gateway to make use of the internal API easier.

Also, to allow programatic access to the API key, a new resource policy
is created to allow the OIDC identity provider enough permissions to
fetch the API key.

It removes the `CoreFrontLocal` parameter in favour of defaulting to
this behavour for dev accounts. This is because changing back to a
private API from the new regional version is a two step process - so not
as easy as just toggling the parameter.

It's a two step process as you can't change from regional to private if
it has custom domains associated with it. In CloudFormation there isn't
a way to specify that something should be removed before making changes,
resulting in having to remove the custom domain before changing the API
to private.

Ideally I would have liked to have conditionally created the UsagePlan -
only for the dev accounts. However there appears to be a bug in cfn-lint
that means we can't use an If function with the CreateUsagePlan
property.

The GHA workflow included in this PR is just to demonstrate that the
gateway is now available with an API key, and also how to
programatically access the API key. This would not be merged with the
PR.

Wynndow committed:
PYIC-5872: Don't use SAM UsagePlan

SAM allows an easy way to create a usage plan and API key, by defining
the `UsagePlan` property in an `Api` resource. Unfortunately you can't
use an `!If` function with it do conditionally create it.

This removes the use of it and defines the resources requied using
plain old CloudFormation. This allows us to only create them when we
want.

Using the previous method would have caused a usage plan and api key to
be created in all envs including prod. This felt like murky ground - the
non dev api's would still be private so it wouldn't have been an issue,
but feels safer to only create them when actually needed.

Wynndow commented on .github/workflows/dev-env-api-test.yaml:
```diff
@@ -0,0 +1,36 @@
+name: Dev env internal API test
```
This is just hear to demonstrate the PR works, and how to access the API key programatically. Will be removed before merging the PR.

Joe-Edwards-GDS commented on deploy/template.yaml:
```diff
@@ -325,18 +310,20 @@ Resources:
           /foo:
-            bar: baz # workaround to get `sam validate` to work
+            bar:
+              baz: thing # workaround to get `sam validate` to work
```
Is this just to make it detect a change?

Wynndow replied:
Good question. sam validate was moaning at me that it wasn't a map. I'm going to try put it back.

Wynndow replied:
Seems happy now ¯\_(ツ)_/¯

Wynndow committed:
PYIC-5872: Create separate API for internal testing

This creates a new API gateway to run in parallel with the original.
This means that the VPC protected internal API can still be used by
core-front without any changes, and the new regional, API token
protected API gateway can be used for testing.

cfn-lint has an issue with having an event defined in a lambda that
includes a reference to a conditionally created resource, even if that
resource is behind the same condition. This has lead to the ignore
written into the metadata of a number of resources.

Joe-Edwards-GDS commented on deploy/template.yaml:
```diff
@@ -583,6 +608,11 @@ Resources:
+      cfn-lint:
+        config:
+          ignore_checks:
+            - W1001 # Ref to resource "IPVCoreInternalTestingApi" that may not be available when condition "IsDevelopment" is False
```
This is a bit sad 😞 - I guess it's not clever enough to realise that the condition is the same

(I assume you've checked that it _isn't_ a problem - not actually sure which order it'll resolve everything!)

Wynndow replied:
I've not deployed to a non-dev environment to test it. But the validation failure breaks our existing workflows which is annoying. I think it's to do with how SAM gets transformed and generates new resources.

Joe-Edwards-GDS reviewed (commented):
This seems better to me, what do you think?

It makes the prod configuration cleaner, and isn't much more code than before. The main thing that's annoying is configuring the extra events and overriding that cfn-lint rule :/

Joe-Edwards-GDS reviewed (commented):
I'm happy to approve this, but I think we need to:
- Remove the test pipeline code
- Monitor this in `build` to make sure it doesn't cause any changes there

Wynndow committed:
PYIC-5872: Remove test workflow

This workflow was purely to demonstrate the new gateway was function and
that the API key could be retrieved.

Removing in a separate commit rather than squashing into the previous
commit to keep a record of the workflow in git in case we want to refer
to it.
